### PR TITLE
Sema: Apply metatype-to-object conversions even without restriction hints.

### PR DESCRIPTION
**Pull request by jckarter**
In some cases, such as when pushing a collection conversion down to per-element conversions, we'll coerce a subtype metatype to AnyObject, as in:

```
func f(_: [AnyObject]) {}
f([NSString.self, NSObject.self]) // Type checks as [NSObject.Type] converted to [AnyObject]
```

and only record the restriction kinds used in the indirect steps NSString -> NSObject and NSObject.Type -> AnyObject without recording the jump from NSString.Type to AnyObject. coerceToType ought to apply this subtyping rule even without such a hint, though, since the restriction kind is intended only as an optimization. Fixes rdar://problem/42666956 .

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/AST/ASTVerifier.cpp [hunk 1]
```diff
@@ -1406,6 +1406,145 @@ class Verifier : public ASTWalker {
...
+        }
+      }
+
+      auto erasedLayout = erasedTy->getCanonicalType()->getExistentialLayout();
+
+      // An existential-to-existential erasure ought to reduce the set of
+      // constraints.
+      if (knownConcreteErasure != ConcreteErasureOnly
+          && concreteTy->isExistentialType()) {
+        // TODO
```
Comment (xedin) [inline lib/AST/ASTVerifier.cpp]: Is it safe to commit this with TODO?

Comment (xedin) [reply lib/AST/ASTVerifier.cpp]: I guess it's in verifier, but still...

**Review by xedin — APPROVED** [review]
LGTM!

**Review by rudkx — APPROVED** [review]
LGTM.

File: lib/AST/ASTVerifier.cpp (hunk 1 above)
Comment (jckarter) [reply lib/AST/ASTVerifier.cpp]: Should be, there was no verification of ErasureExprs at all before this patch. I was going to finish that up as a followup.

**Pull request merged**